traffic_router/core/src/main/java/com/comcast/cdn/traffic_control/traffic_router/core/router/TrafficRouter.java

```diff
@@ -560,16 +564,50 @@ public HTTPRouteResult multiRoute(final HTTPRequest request, final Track track)
 		return routeResult;
 	}
 
+	/**
+	 * Creates a string to be used in consistent hashing
+	 *
+	 * This uses simply the request path by default, but will consider any and all Query Parameters
+	 * that are in deliveryService's {@link DeliveryService.consistentHashQueryParams} set as well.
+	 * It will also fall back on the request path if the query parameters are not UTF-8-encoded.
+	 *
+	 * @param deliveryService The Delivery Service being requested
+	 * @param request An {@link HTTPRequest} representing the client's request.
+	 * @return A string appropriate to use for consistent hashing to service the request
+	*/
+	@SuppressWarnings({"PMD.CyclomaticComplexity", "PMD.UseStringBufferForStringAppends"})
 	public String buildPatternBasedHashString(final DeliveryService deliveryService, final HTTPRequest request) {
 		final String requestPath = request.getPath();
 		if (deliveryService.getConsistentHashRegex() != null && !deliveryService.getConsistentHashRegex().isEmpty() && !requestPath.isEmpty()) {
 			final StringBuilder hashString = new StringBuilder(request.getPath());
-			for (String qkey : request.getQueryString().split("&")) {
-				qkey = qkey.split("=")[0];
-				if (deliveryService.consistentHashQueryParams.contains(qkey)) {
-					hashString.append(qkey);
+			final SortedSet<String> qparams = new TreeSet<String>();
+			try {
+				for (final String qparam : request.getQueryString().split("&")) {
+					if (qparam.isEmpty()) {
+						continue;
+					}
+
+					String[] parts = qparam.split("=");
+					for (short i = 0; i<parts.length; ++i) {
+						parts[i] = URLDecoder.decode(parts[i]);
+					}
+					parts[0] = parts[0].toLowerCase(); // I truly have no idea why this triggers `PMD.UseStringBufferForStringAppends`
```

Comment:
I'm not sure we should be lowercasing the input. That would make it case-insensitive, and I'm not sure query strings are generally case-insensitive.

Reply:
I'm pretty sure the database constraint on uniqueness is case-insensitive (or at least it was at one point - maybe that's changed?) so this is necessary for consistency.

Reply:
After discussion, the db constraint _was_ case-insensitive, but isn't anymore. should be changed in my latest.